guppylang/checker/cfg_checker.py

```diff
@@ -260,13 +266,43 @@ def check_bb(
     return checked_bb
 
 
-def check_rows_match(row1: Row[Variable], row2: Row[Variable], bb: BB) -> None:
+def check_rows_match(
+    row1: Row[Variable], row2: Row[Variable], bb: BB, globals: Globals
+) -> None:
     """Checks that the types of two rows match up.
 
     Otherwise, an error is thrown, alerting the user that a variable has different
     types on different control-flow paths.
     """
     map1, map2 = {v.name: v for v in row1}, {v.name: v for v in row2}
+
+    # If block signature lengths don't match but no undefined error was thrown, some
+    # variables may be shadowing global variables.
+    if map1.keys() != map2.keys():
+        for x in map1.keys() ^ map2.keys():
+            if x in globals:
+                v = map1.get(x) or map2.get(x)
+                assert v is not None
+                g = globals[x]
+                assert hasattr(g, "ty")
+                if v.ty != g.ty:
+                    ident = (
+                        "Expression"
+                        if v.name.startswith("%")
+                        else f"Variable `{v.name}`"
+                    )
+                    use = bb.containing_cfg.live_before[bb][v.name].vars.used[v.name]
+                    err = BranchTypeError(use, ident)
+                    err.add_sub_diagnostic(BranchTypeError.TypeHint(v.defined_at, v.ty))
+                    # We don't add a location to the type hint for the global variable,
+                    # since  it could lead to cross-file diagnostics (which are not
+                    # supported) or refer to long function definitions.
```

Comment:
Imo this would be a good reason to support cross-file diagnostics, but that is a separate issue 😅